storage/jellyfish-merkle/src/lib.rs

```diff
@@ -392,55 +365,128 @@ where
     /// the returned batch, the state `S_{i+1}` is ready to be read from the tree by calling
     /// [`get_with_proof`](struct.JellyfishMerkleTree.html#method.get_with_proof). Anything inside
     /// the batch is not reachable from public interfaces before being committed.
-    pub fn batch_put_value_set(
+    ///
+    /// Assumes 16 shards in total here.
+    pub fn batch_put_value_set_for_shard(
         &self,
+        shard_id: u8,
         value_set: Vec<(HashValue, Option<&(HashValue, K)>)>,
         node_hashes: Option<&HashMap<NibblePath, HashValue>>,
         persisted_version: Option<Version>,
         version: Version,
-    ) -> Result<(HashValue, TreeUpdateBatch<K>)> {
+    ) -> Result<(Node<K>, TreeUpdateBatch<K>)> {
         let deduped_and_sorted_kvs = value_set
             .into_iter()
+            .map(|kv| {
+                assert!(kv.0.nibble(0) == shard_id);
+                kv
+            })
             .collect::<BTreeMap<_, _>>()
             .into_iter()
             .collect::<Vec<_>>();
 
-        let mut batch = TreeUpdateBatch::new();
-        let root_node_opt = if let Some(persisted_version) = persisted_version {
+        let shard_root_node_key = NodeKey::new(version, NibblePath::new_odd(vec![shard_id << 4]));
```

Comment:
provide some comment for `NibblePath::new_odd(vec![shard_id << 4])`? hard to understand without comment

Reply:
Done